sos/collector/transports/saltstack.py

```diff
@@ -0,0 +1,175 @@
+        return True
+
+    def _connect(self, password=None):    # sourcery skip: replace-interpolation-with-fstring
+        """Connect to the remote host using SaltStack Master.
+
+        This method will attempt to connect to the remote host using SaltStack
+        Master. If the connection fails, an exception will be raised.
+
+        If the connection is successful, the connection will be stored in the
+        self._connection attribute.
+        """
+        if not is_executable('salt'):
+            return False
+
+        try:
+            self._check_for_saltstack()
+        except ConnectionException:
+            self.log_error("Transport is not locally supported. ")
+            raise
+        self.log_info("Transport is locally supported and service running. ")
+        cmd = "test.ping"
+        result = self.run_salt(cmd, timeout=180)
+        return result['status'] == 0
+
+    def _disconnect(self):
+        return True
+
+    @property
+    def remote_exec(self):
+        return ("salt --out json --static %s cmd.shell " % (self.address))
+
+    def _retrieve_file(self, fname, dest):
+        # sourcery skip: replace-interpolation-with-fstring
+        """Retrieve a file from the remote host using saltstack cp.push
+
+        This method will retrieve a file from the remote host and access it
+        from the master's minion cachedir.
+
+        Parameters
+            fname       The path to the file on the remote host
+            dest        The path to the destination directory on the master
+
+        Returns
+            True if the file was retrieved, else False
+        """
+        if not self.connected:
+            return False
+        if not os.path.exists(dest):
+            os.makedirs(dest)
+        master_minion_cachedir = "/var/cache/salt/master/minions/%s/files"
+        master_minion_cachedir %= self.address
+        cmd = "cp.push %s remove_source=true" % fname
+        result = self.run_salt(cmd, timeout=180)
+        if result['status'] == 0:
+            self.log_info("File %s retrieved from minion" % fname)
+            shutil.copy("%s/%s" % (master_minion_cachedir, fname), dest)
+            return True
+        return False
+
+    def _get_hostname(self):
+        """Determine the hostname of the node and set that for future reference
+        and logging
+
+        :returns:   The hostname of the system, per the `hostname` command
+        :rtype:     ``str``
+        """
+        _out = self.run_command('hostname')
+        if _out['status'] == 0:
+            self._hostname = list(json.loads(_out['output']).values())[0].strip()
+
+        if not self._hostname:
+            self._hostname = self.address
+        self.log_info("Hostname set to %s" % self._hostname)
+        return self._hostname
+
+    def _read_file(self, fname):
+        res = self.run_command("cat %s" % fname, timeout=10)
+        output = list(json.loads(res['output']).values())[0]
+        output += output + '\n'
```

Comment:
Good catch, not root cause. This was added this at the very last minute last night before calling it quits. For some reason **control_persist** transport always shows a blank line in `/etc/os-release` even when the file has no blank line. The **saltstack** transport always shows the file exactly, no extra blank line. This was a hack to just try to confirm the blank line at the end didn't fix finding the RockyLinux policy from the data returned in `os-release`.

I'll remove this shortly since an extra blank line did not change the behavior.

Reply:
By "shows" I mean if I just added `print(output)` to any `_read_file()` to examine the data obtained by the transport.

Reply:
> If we can drop --out json then it looks like we should also be able to use the base _read_file() as well.

I didn't look deep enough into collect & transport code to confirm all possible regex patterns and if salt modifying the data would cause failures. I will gladly use the default functions if you think this is an unlikely issue.